src/main/java/org/jitsi/videobridge/cc/vp8/VP8Frame.java

```diff
@@ -53,131 +49,158 @@
+     * A boolean that indicates whether the incoming VP8 frame that this
+     * instance refers to is a reference frame (RFC7741).
      */
-    private final int maxSequenceNumberSeenBeforeFirstPacket;
+    private final boolean isReference;
 
     /**
-     * The max sequence number that will be forwarded for this frame. Note that
-     * this is not necessarily the ending sequence number.
+     * A record of how this frame was projected, or not.
      */
-    private int maxSequenceNumber = -1;
+    private VP8ProjectionRecord projection;
 
     /**
      * Ctor.
      *
-     * @param firstPacketOfFrame The first VP8 packet of the frame that the
-     * packet refers to.
-     * @param maxSequenceNumberSeenBeforeFirstPacket the max sequence number
-     * that was seen before the arrival of the first packet of this frame. This
-     * is useful for piggybacking any mis-ordered packets.
+     * @param packet A packet from the frame to be constructed.
      */
-    VP8Frame(@NotNull VideoRtpPacket firstPacketOfFrame,
-             int maxSequenceNumberSeenBeforeFirstPacket)
+    VP8Frame(@NotNull Vp8Packet packet)
     {
-        this.ssrc = firstPacketOfFrame.getSsrc();
-        this.timestamp = firstPacketOfFrame.getTimestamp();
-        this.startingSequenceNumber = firstPacketOfFrame.getSequenceNumber();
-        this.maxSequenceNumberSeenBeforeFirstPacket
-            = maxSequenceNumberSeenBeforeFirstPacket;
-
-        byte[] buf = firstPacketOfFrame.getBuffer();
-        int payloadOffset = firstPacketOfFrame.getPayloadOffset(),
-            payloadLen = firstPacketOfFrame.getPayloadLength();
-
-        this.tl0PICIDX = DePacketizer
-            .VP8PayloadDescriptor.getTL0PICIDX(buf, payloadOffset, payloadLen);
-        this.isKeyframe = DePacketizer
-            .isKeyFrame(buf, payloadOffset, payloadLen);
-        this.isReference = DePacketizer
-            .VP8PayloadDescriptor.isReference(buf, payloadOffset, payloadLen);
-        this.isTL0 = DePacketizer.VP8PayloadDescriptor
-            .getTemporalLayerIndex(buf, payloadOffset, payloadLen) == 0;
+        this.ssrc = packet.getSsrc();
+        this.timestamp = packet.getTimestamp();
+        this.earliestKnownSequenceNumber = packet.getSequenceNumber();
+        this.latestKnownSequenceNumber = packet.getSequenceNumber();
+        this.seenStartOfFrame = packet.isStartOfFrame();
+        this.seenEndOfFrame = packet.isEndOfFrame();
+        this.numPacketsSeen = 1;
+
+        byte[] buf = packet.getBuffer();
+        int payloadOffset = packet.getPayloadOffset(),
+            payloadLen = packet.getPayloadLength();
+
+        this.tl0PICIDX = packet.getTL0PICIDX();
+        this.isKeyframe = packet.isKeyframe();
+        this.pictureId = packet.getPictureId();
+        this.isReference =
+            /* TODO add this to Vp8Packet if we need it, otherwise remove it. */
+            DePacketizer.VP8PayloadDescriptor.isReference(buf, payloadOffset, payloadLen);
```

Comment:
Guess we should add a helper on `Vp8Packet` for this.  Probably only `Vp8Packet` should use `DePacketizer` and expose nice methods for other places to use.

Reply:
(not saying in this PR, just a thought)

Reply:
Yeah, that's what I had in mind with that TODO.  If I even need `isReference`, which so far I haven't needed to use.

Reply:
🤦‍♂ sorry, managed to completely miss that comment.